### ISSUE TITLE
Ability to save window layout and persist between sessions

### ISSUE THREAD

**Issue by robbyoconnor** [syl20bnr/spacemacs#2832]
It would be super neat if we could save our window layout and persist it between sessions. Making an issue just in case anyone wants it...also to open a dialogue :)

**Comment by TheBB** [comment]
I think this is essentially `desktop-save-mode`? Some people have tried to get it working, but it seems more difficult than normal, for some reason.

#1265 #1112

**Comment by robbyoconnor** [comment]
Ah -- I tried to a search -- but I guess my keywords sucked...

**Comment by usharf** [comment]
There's seem there's some setup for this in spacemacs' config.el, which is disabled...

**Comment by robbyoconnor** [comment]
so it seems...I guess having this issue isn't going to harm things...I don't know why it's commented out...probably because it doesn't play well...and breaks things...

**Comment by bmag** [comment]
Well, config.el says about the desktop configuration that "The following code is kept as reference" (line 281).
There is also some minimal config in packages.el (lines 453-459) that just makes sure the desktop file is saved in the cache directory.

I have used desktop-save-mode several times manually - `desktop-save` before exiting spacemacs, then `desktop-read` after re-opening spacemacs. I didn't encounter any problems, and the window layout was indeed restored. However, I don't enable `desktop-save-mode` automatically in my config, so I can't say about using it full time.

**Comment by robbyoconnor** [comment]
@bmag was it in spacemacs?

**Comment by bmag** [comment]
yes

**Comment by robbyoconnor** [comment]
This comment is strictly for reference, once #1958 is merged into `develop`, I think this can be safely closed.

**Comment by Devagamster** [comment]
I bound some custom bindings to desktop-save-in-desktop-dir and desktop-read which I manually run whenever I need it. In my experience these have worked fine, but the automatic saving and loading does not seem to work from desktop-save-mode. No error, but it doesn't reload the window layout.

**Comment by syl20bnr** [comment]
@bmag would be cool to be able to save multiple sessions and listing them on the home buffer.

**Comment by bmag** [comment]
@syl20bnr if you're talking about sessions as understood by the `desktop` package (i.e. also restore buffers), then I don't know if it's possible, but it might be. If you're talking about `window-purpose`, then we should talk about window layouts (without restoring buffers), and yes it can be done.

**Comment by StreakyCobra** [comment]
`spacemacs-layout` is now in `develop` :relaxed:

**Comment by tuhdo** [comment]
[persp-mode](https://github.com/Bad-ptr/persp-mode.el) is an alternative version of [perspective](https://github.com/nex3/perspective-el). It can do everything `perspective` does, plus saving your workspaces and the layouts in them across session. When you open Emacs again, previous workspace with its layout and visible buffers are loaded accordingly.

**Comment by tuhdo** [comment]
Here is my configuration for `persp-mode`:

```lisp
(defun tuhdo/init-persp-mode ()
  (use-package persp-mode
    :init
    (setq persp-save-dir spacemacs-cache-directory)
    (setq persp-lighter "")
    (evil-leader/set-key
      "Ll"  'persp-load-state-from-file
      "Ls"  'persp-save-state-to-file
      "La"  'persp-add-buffer
      "LA"  'persp-set-buffer
      "Lc"  'persp-kill
      "Lk"  'persp-remove-buffer
      "Ln"  'persp-next
      "Lp"  'persp-prev
      "Lr"  'persp-rename
      "Ls"  'persp-switch)
    (with-eval-after-load 'persp-mode
      (with-eval-after-load 'window-numbering
        (define-key window-numbering-keymap (kbd "M-1") 'persp-add-buffer)
        (define-key window-numbering-keymap (kbd "M-2") 'persp-remove-buffer)
        (define-key window-numbering-keymap (kbd "M-3") 'persp-prev)
        (define-key window-numbering-keymap (kbd "M-4") 'persp-next)
        (define-key window-numbering-keymap (kbd "M-5") 'persp-switch))
      (define-key persp-mode-map (kbd "M-1") 'persp-prev)
      (define-key persp-mode-map (kbd "M-2") 'persp-next))))
```

**Comment by tuhdo** [comment]
I've just realized that Spacemacs already included `persp-mode` in `develop`. Woot. I'm so outdated with Spacemacs development already.

**Comment by syl20bnr** [comment]
@thudo `spacemacs layouts` are persp-mode perspectives, there are a couple of new dotfile variable for layouts, also check the README of the spacemacs layout layer (which is included in the spacemacs distribution).

**Comment by mooreniemi** [comment]
@tuhdo thanks for sharing your config. my only tweak was to change the mapping for switch, as it was shadowing save.
